python/lsst/ts/hvac/xml/hvac_mqtt_to_SAL_XML.py

```diff
@@ -19,153 +19,495 @@
-    global commands
+    Returns
+    -------
+    topic: `str`
+        Defined as the part before the last forward slash.
+    item: `str`:
+         Defined as the part after the last forward slash.
+
+    Raises
+    ------
+    ValueError
+        In case no forward slash is found.
+    """
+    # This throws a ValueError in case no forward slash is found.
+    topic_end = topic_and_item.rindex("/")
+    # omit the forward slash
+    item_start = topic_end + 1
+    topic = topic_and_item[:topic_end]
+    item = topic_and_item[item_start:]
+    return topic, item
+
+
+def get_topics():
+    """Convenience method to collect all generic topics, representing the HVAC
+    subsystems.
+
+    Returns
+    -------
+    topics: `set`
+        A set of all generic topics.
+
+    """
+    global hvac_topics
+    if not bool(hvac_topics):
+        collect_topics_and_items()
+    topics = set()
+    for topic_and_item in hvac_topics:
+        topic_type = hvac_topics[topic_and_item]["topic_type"]
+        if topic_type == "WRITE" and topic_and_item.endswith("COMANDO_ENCENDIDO_LSST"):
+            topic, item = extract_topic_and_item(topic_and_item)
+            topics.add(topic)
+    for topic in TOPICS_ALWAYS_ENABLED:
+        topics.add(topic)
+
+    return topics
+
+
+def get_telemetry_mqtt_topics_and_items():
+    """Convenience method to collect all MQTT topics and their items that
+    publish telemetry.
+
+    Returns
+    -------
+    telemetry_mqtt_topcis: `dict`
+        A dictionary with the MQTT topics as keys and their items as value.
+
+    Notes
+    -----
+    The structure of the returned dictionary is exactly the same as for the
+    hvac_topics dictionary. The only difference is that the command topics
+    (indicated by topic type WRITE) have been filtered out.
+    """
+    telemetry_mqtt_topcis = {}
+    for hvac_topic in get_topics():
+        items = get_items_for_topic(hvac_topic)
+        topic_items = {}
+        for item in items:
+            # Make sure only telemetry topics are used.
+            if items[item]["topic_type"] == "READ":
+                topic_items[item] = items[item]
+        telemetry_mqtt_topcis[hvac_topic] = topic_items
+    return telemetry_mqtt_topcis
+
+
+def _parse_limits(limits_string):
+    """Parse the string value of the limits column by comparing it to known
+    regular expressions and extracting the minimum and maximum values.
+
+    Parameters
+    ----------
+    limits_string: `str`
+        The string containing the limits to parse.
+
+    Returns
+    -------
+    lower_limit: `float`
+        The lower limit
+    upper_limit: `float`
+        The upper limit
+
+    Raises
+    ------
+    ValueError
+        In case an unknown string pattern is found in the limits column.
+
+    """
+    lower_limit = -99.9
+    upper_limit = 99.9
+    matched = False
+    for pattern in [
+        r"^(-*\d+)/(-*\d+)$",
+        r"^(-*\d+) a (-*\d+)$",
+        r"^(-*\d+)% a (-*\d+)%$",
+    ]:
```

Comment:
If this is run frequently then please consider compiling your regex strings once somewhere, then iterating over the list of compiled strings. It's supposed to be faster but is a bit of a nuisance to have to keep track of them.

Also, please consider a single pattern, such as the following:
```
r"^(-*\d+)\%? */|a *(-*\d+)\%?$"
```
It's a bit more permissive. You can easily be just as picky about the middle (requiring "/" or " a "), but it's significantly more work to require both % or neither % -- though it can be done with a backreference.

Reply:
Good point about the pattern matching. I rewrote it to a single pattern which is slightly different than what you suggested. I also rewrote the if statements so I could get rid of the boolean that keeps track of whether a match was made or not.